File: packages/haskell/src/ArrayFirst.hs
-- | Get the first element or first n elements of a list.
--   Haskell port of https://github.com/mattstrick/array-first-javascript
module ArrayFirst
  ( arrayFirst,
    arrayFirstN,
  )
where

-- | Returns the first element, or Nothing if the list is empty.
arrayFirst :: [a] -> Maybe a
arrayFirst [] = Nothing
arrayFirst (x : _) = Just x

-- | Returns the first n elements. If n <= 0, treated as 1.
--   When n is 1, returns a single-element list.
arrayFirstN :: [a] -> Int -> Maybe [a]
arrayFirstN [] _ = Nothing
arrayFirstN xs n =
  let k = max 1 n
   in Just (take k xs)

File: packages/elixir/lib/array_first.ex
defmodule ArrayFirst do
  @moduledoc """
  Get the first element or first n elements of a list.
  Elixir port of https://github.com/mattstrick/array-first-javascript
  """

  @doc """
  Returns the first element or first n elements of a list.

  - `first(list)` — returns the first element, or `nil` if empty.
  - `first(list, n)` — returns the first n elements (single element if n is 1).
  """
  def first(list) when is_list(list) do
    case list do
      [] -> nil
      [h | _] -> h
    end
  end

  def first(list, n) when is_list(list) do
    n = if is_number(n) and n >= 1, do: trunc(n), else: 1
    case list do
      [] -> nil
      _ when n == 1 -> hd(list)
      _ -> list |> Enum.take(n)
    end
  end
end
